core/coreapi/pubsub.go

```diff
@@ -0,0 +1,161 @@
+package coreapi
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"sync"
+	"time"
+
+	core "github.com/ipfs/go-ipfs/core"
+	coreiface "github.com/ipfs/go-ipfs/core/coreapi/interface"
+	caopts "github.com/ipfs/go-ipfs/core/coreapi/interface/options"
+
+	cid "gx/ipfs/QmPSQnBKM9g7BaUcZCvswUJVscQ1ipjmwxN5PXCjkp9EQ7/go-cid"
+	floodsub "gx/ipfs/QmPbYVFhKxamZAN9MyrQMDeoGYa6zkQkhAPguwFfzxPM1J/go-libp2p-floodsub"
+	peer "gx/ipfs/QmbNepETomvmXfz1X5pHNFD2QuPqnqi47dTd94QJWSorQ3/go-libp2p-peer"
+	pstore "gx/ipfs/QmfAQMFpgDU2U4BXG64qVr8HSiictfWvkSBz7Y2oDj65st/go-libp2p-peerstore"
+)
+
+type PubSubAPI CoreAPI
+
+type pubSubSubscription struct {
+	subscription *floodsub.Subscription
+}
+
+type pubSubMessage struct {
+	msg *floodsub.Message
+}
+
+func (api *PubSubAPI) Ls(ctx context.Context) ([]string, error) {
+	if err := api.checkNode(); err != nil {
+		return nil, err
+	}
+
+	return api.node.Floodsub.GetTopics(), nil
+}
+
+func (api *PubSubAPI) Peers(ctx context.Context, opts ...caopts.PubSubPeersOption) ([]peer.ID, error) {
+	if err := api.checkNode(); err != nil {
+		return nil, err
+	}
+
+	settings, err := caopts.PubSubPeersOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	peers := api.node.Floodsub.ListPeers(settings.Topic)
+	out := make([]peer.ID, len(peers))
+
+	for i, peer := range peers {
+		out[i] = peer
+	}
+
+	return out, nil
+}
+
+func (api *PubSubAPI) Publish(ctx context.Context, topic string, data []byte) error {
+	if err := api.checkNode(); err != nil {
+		return err
+	}
+
+	return api.node.Floodsub.Publish(topic, data)
+}
+
+func (api *PubSubAPI) Subscribe(ctx context.Context, topic string, opts ...caopts.PubSubSubscribeOption) (coreiface.PubSubSubscription, error) {
+	options, err := caopts.PubSubSubscribeOptions(opts...)
+
+	if err := api.checkNode(); err != nil {
+		return nil, err
+	}
+
+	sub, err := api.node.Floodsub.Subscribe(topic)
+	if err != nil {
+		return nil, err
+	}
+
+	if options.Discover {
+		go func() {
+			blk, err := api.core().Block().Put(ctx, strings.NewReader("floodsub:"+topic))
```

Comment:
This use of the context is a bit awkward. I'd expect the context to only apply to the lifetime of the `Subscribe` operation but it really applies until we're fully bootstrapped.

Maybe we should:

1. Create an entirely new (independent) context.
2. Cancel it when the subscription is closed.

(in that case, `Subscribe` may not need to take a context but it might be a good idea to leave it in the signature just in case)